### PR TITLE
Remove legacy code from System.Text.Encodings.Web

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
This is preparing the __System.Text.Encodings.Web__ project so that we can resolve https://github.com/dotnet/corefx/issues/34830 and https://github.com/dotnet/corefx/issues/33509.

Currently this PR does the following things:

- Removes almost all _unsafe_ code from the implementations, relying wholly on `Rune` and `Span` as workhorse types.
- Removes legacy APIs that don't make sense to carry forward and that we can't make work in the new world.
- Rewrites the base type `TextEncoder` on top of only two abstract methods: does this scalar value have to be encoded, and if so what does the encoding look like? Anybody can write a `TextEncoder` instance by subclassing the type and overriding only those two methods.

A future iteration will:

- Improve the performance of these types, including vectorizing code where possible.
- Add UTF-8 overloads, `Span`-based overloads, and `OperationStatus`-based overloads of the `Encode` routines.
- Move the type in-box so that it can be used by __System.Text.Json__.

This PR in its current form is not attempting to optimize the performance of these routines. I'm concerned only with the API shape at the moment. Performance improvements can come in a future PR.

__Note to reviewers:__ These APIs are __not__ part of net standard, and they're __not__ currently shipping in-box. Currently the only way to get this package is to perform a manual install from NuGet. (ASP.NET 2.x also includes the package in its project template.)

### CODE REVIEW

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs [hunk 1]
```diff
@@ -79,89 +79,70 @@ internal static void ForbidHtmlCharacters(AllowedCharactersBitmap allowedCharact
...
-        public unsafe override int FindFirstCharacterToEncode(char* text, int textLength)
-        {
-            return _allowedCharacters.FindFirstCharacterToEncode(text, textLength);
-        }
-
-        public override int MaxOutputCharactersPerInputCharacter
+        public override bool RuneMustBeEncoded(Rune value)
         {
-            get { return 10; } // "&#x10FFFF;" is the longest encoded form
+            return !_allowedCharacters.IsUnicodeScalarAllowed((uint)value.Value);
```
Comment (tarekgh) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs]: >(uint) [](start = 62, length = 6)

Why did you change the IsUnicodeScalarAllowed parameter from int to uint if we are going to cast it again in the calling side?

**Comment by tarekgh** [comment]
The CI is failing with 

```
D:\a\1\s\.packages\microsoft.dotnet.build.tasks.packaging\1.0.0-beta.19216.2\build\Packaging.targets(1141,5): error : System.Text.Encodings.Web should be supported on .NETCoreApp,Version=v1.0/win7-x86 but has no compile assets. [D:\a\1\s\src\System.Text.Encodings.Web\pkg\System.Text.Encodings.Web.pkgproj]
D:\a\1\s\.packages\microsoft.dotnet.build.tasks.packaging\1.0.0-beta.19216.2\build\Packaging.targets(1141,5): error : System.Text.Encodings.Web should be supported on .NETCoreApp,Version=v1.0/win7-x64 but has no compile assets. [D:\a\1\s\src\System.Text.Encodings.Web\pkg\System.Text.Encodings.Web.pkgproj]
D:\a\1\s\.packages\microsoft.dotnet.build.tasks.packaging\1.0.0-beta.19216.2\build\Packaging.targets(1141,5): error : System.Text.Encodings.Web should be supported on .NETCoreApp,Version=v1.0/osx.10.11-x64 but has no compile assets. [D:\a\1\s\src\System.Text.Encodings.Web\pkg\System.Text.Encodings.Web.pkgproj]
D:\a\1\s\.packages\microsoft.dotnet.build.tasks.packaging\1.0.0-beta.19216.2\build\Packaging.targets(1141,5): error : System.Text.Encodings.Web should be supported on .NETCoreApp,Version=v1.0/centos.7-x64 but has no compile assets. [D:\a\1\s\src\System.Text.Encodings.Web\pkg\System.Text.Encodings.Web.pkgproj]
D:\a\1\s\.packages\microsoft.dotnet.build.tasks.packaging\1.0.0-beta.19216.2\build\Packaging.targets(1141,5): error : System.Text.Encodings.Web should be supported on .NETCoreApp,Version=v1.0/debian.8-x64 but has no compile assets. [D:\a\1\s\src\System.Text.Encodings.Web\pkg\System.Text.Encodings.Web.pkgproj]
```

**Comment by blowdart** [comment]
Let me channel fowler; performance data (because this will reflect in asp.net)

**Comment by GrabYourPitchforks** [comment]
My gut tells me that this will have a small but hopefully insignificant perf impact due to the use of `Span` everywhere instead of raw pointers. The highest priority at the moment is adjusting the API shape to be future-proof. Performance optimizations can come later as an implementation detail.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AllowedCharactersBitmap.cs [hunk 2]
```diff
@@ -70,32 +70,21 @@ public AllowedCharactersBitmap Clone()
...
-            for (int i = 0; i < textLength; i++)
+            uint index = unicodeScalar >> 5;
+            if (index >= (uint)_allowedCharacters.Length)
             {
-                if (!IsCharacterAllowed(text[i])) { return i; }
+                return false; // would be out-of-bounds read (unicodeScalar isn't BMP char)
             }
-            return -1;
+
+            return ((_allowedCharacters[index] >> ((int)unicodeScalar & 0x1F) /* offset */) & 1U) != 0;
```
Comment (krwq) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AllowedCharactersBitmap.cs]: presumably _allowedCharacters only stores info about `2^21` characters but unicodeScalar is 32 bit, is there any chance this can throw IndexOutOfRange?

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs [hunk 3]
```diff
@@ -79,89 +79,70 @@ internal static void ForbidHtmlCharacters(AllowedCharactersBitmap allowedCharact
...
-                *buffer = HexUtil.Int32LsbToHexDigit(unicodeScalar & 0xF);
-                unicodeScalar >>= nibbleSize;
-                buffer--;
-            }
-            while (unicodeScalar != 0);
-
-            buffer += numberOfHexCharacters + 1;
-            *buffer = ';';
-            return true;
+                buffer[--offset] = HexUtil.UInt32LsbToHexDigit(scalarValue & 0xF);
```
Comment (krwq) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs]: `Debug.Assert(offset >= 0)`?

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AllowedCharactersBitmap.cs (hunk 2 above)
Comment (GrabYourPitchforks) [reply src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AllowedCharactersBitmap.cs]: `_allowedChars` is 8 KB (a bitmap holding 65,536 bits). We now perform a length check at the beginning of this method before accessing the array, so it should never go out of bounds.

The original implementation never would have gone out of bounds anyway (since the caller doesn't pass values outside the valid range), but adding the check is good practice nonetheless and it results in slightly more optimized codegen.

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test this please

**Comment by GrabYourPitchforks** [comment]
Turns out that this is untenable for various reasons. Closing and will think about this more later.
